src/projections/adams.cpp

```diff
@@ -0,0 +1,217 @@
+ * and where series evaluation  must be multiplied by the argument. */
+
+    int i = ORDER;
+    double d1 = 0.0;
+    double d2 = 0.0;
+    const double C[] = { /* even coefficients */
+        2.19174570831038,
+        0.0914203033408211,
+        -0.00575574836830288,
+        -0.0012804644680613,
+        5.30394739921063e-05,
+        3.12960480765314e-05,
+        2.02692115653689e-07,
+        -8.58691003636495e-07};
+    double const *Cp = C + ORDER - 1;
+
+    double y = phi * M_2_PI;
+    y = 2. * y * y - 1.;
+    double y2 = 2. * y;
+    while (--i) {
+        double temp = d1;
+        d1 = y2 * d1 - d2 + *Cp--;
+        d2 = temp;
+    }
+
+    return phi * (y * d1 - d2 + 0.5 * *Cp);
+}
+
+static PJ_XY forward(PJ_LP lp, PJ *P) {
+    double m, n, a=0., b=0.;
+    int sm=0, sn=0;
+    PJ_XY xy;
+    struct pj_opaque *Q = static_cast<struct pj_opaque*>(P->opaque);
+
+    switch (Q->mode) {
+    case GUYOU:
+        if ((fabs(lp.lam) - TOL) > M_PI_2) {
+            proj_errno_set(P, PJD_ERR_TOLERANCE_CONDITION);
+        }
+
+        if (fabs(fabs(lp.phi) - M_PI_2) < TOL) {
+            xy.x = 0;
+            xy.y = lp.phi < 0 ? -1.85407 : 1.85407;
+            return xy;
+        } else {
+            double sl = sin(lp.lam);
+            double sp = sin(lp.phi);
+            double cp = cos(lp.phi);
+            a = aacos(P->ctx, (cp * sl - sp) * RSQRT2);
+            b = aacos(P->ctx, (cp * sl + sp) * RSQRT2);
+            sm = lp.lam < 0.;
+            sn = lp.phi < 0.;
+        }
+        break;
+    case PEIRCE_Q: {
+            double sl = sin(lp.lam);
+            double cl = cos(lp.lam);
+            double cp = cos(lp.phi);
+            a = aacos(P->ctx, cp * (sl + cl) * RSQRT2);
+            b = aacos(P->ctx, cp * (sl - cl) * RSQRT2);
+            sm = sl < 0.;
+            sn = cl > 0.;
+        }
+        break;
+    case ADAMS_HEMI: {
+            double sp = sin(lp.phi);
+            if ((fabs(lp.lam) - TOL) > M_PI_2) {
+                proj_errno_set(P, PJD_ERR_TOLERANCE_CONDITION);
+            }
+            a = cos(lp.phi) * sin(lp.lam);
+            sm = (sp + a) < 0.;
+            sn = (sp - a) < 0.;
+            a = aacos(P->ctx, a);
+            b = M_PI_2 - lp.phi;
+        }
+        break;
+    case ADAMS_WSI: {
+            double sp = tan(0.5 * lp.phi);
+            b = cos(aasin(P->ctx, sp)) * sin(0.5 * lp.lam);
+            a = aacos(P->ctx, (b - sp) * RSQRT2);
+            b = aacos(P->ctx, (b + sp) * RSQRT2);
+            sm = lp.lam < 0.;
+            sn = lp.phi < 0.;
+        }
+        break;
+    case ADAMS_WSII: {
+            double spp = tan(0.5 * lp.phi);
+            a = cos(aasin(P->ctx, spp)) * sin(0.5 * lp.lam);
+            sm = (spp + a) < 0.;
+            sn = (spp - a) < 0.;
+            b = aacos(P->ctx, spp);
+            a = aacos(P->ctx, a);
+        }
+        break;
+    }
+
+    m = aasin(P->ctx, sqrt(fabs(1. + cos(a + b))));
```

Comment:
the fabs() seems unnecessary here. Unless due to rare floating point issue 1+cos() would be negative ?

Reply:
At least it seems to be unnecessary. I don't know about floating point weirdness. Once I've got successful CI builds I'll try to remove fabs() here and see if it makes a difference.

Reply:
Removing fabs works locally at least